Improved eclipse code editing support

francipvb opened the pull request:
Hi,

It's a patch to solve Eclipse support (Fixes #5667)

Cheers,

francipvb committed:
Improved eclipse code editing support.

francipvb committed:
Cleaning up some code

derekriemer commented:
The only major thing I see here besides maybe being able to extract a little bit of boilerplate code out is this.
+			if obj.role == controlTypes.ROLE_LIST and obj.parent.next.firstChild.name == COMPLETION_DIALOG_LABEL:
You need to not match on the name here, as this is almost certainly a localized string and each language will have a different one. We can't easily have translators translate this because we would need to know exactly what the name is in their language. I propose looking for different ways to detect this window. Does it have IA2UniqueId's or a unique window hwnd or class? If not, can you use parent child relationships in an interesting way? If not, maybe use the role of at least two objects, or see if somehow it fires an event.
You also probably should match on a slightly different attribute for this code as well. +			if obj.role == controlTypes.ROLE_LISTITEM and obj.parent.role == controlTypes.ROLE_LIST and obj.parent.name == AUTOCOMPLETION_LIST_NAME:

One other minor thing I see:
+		except: pass

I would change to  this. I am writing this in diff format, where + indicates a new insertion to the file and - represents a deletion.
-		except: pass
+		except:
+			pass

In other words, break this out on multiple lines to make it consistent with python best practices.
Great work.

francipvb commented:
Hello,

El 14/01/2016 a las 03:26 p.m., derekriemer escribió:
> The only major thing I see here besides maybe being able to extract a
> little bit of boilerplate code out is this.
>
>   * if obj.role == controlTypes.ROLE_LIST and
>     obj.parent.next.firstChild.name == COMPLETION_DIALOG_LABEL: You need
>     to not match on the name here, as this is almost certainly a
>     localized string and each language will have a different one. We
>     can't easily have translators translate this because we would need
>     to know exactly what the name is in their language. I propose
>     looking for different ways to detect this window. Does it have
>     IA2UniqueId's or a unique window hwnd or class? If not, can you use
>     parent child relationships in an interesting way? If not, maybe use
>     the role of at least two objects, or see if somehow it fires an
>     event. You also probably should match on a slightly different
>     attribute for this code as well. + if obj.role ==
>     controlTypes.ROLE_LISTITEM and obj.parent.role ==
>     controlTypes.ROLE_LIST and obj.parent.name == AUTOCOMPLETION_LIST_NAME:

Yes, I'll try to solve it. I'll see if I can find these values.

> One other minor thing I see:
>
>   * except: pass
>
> I would change to this. I am writing this in diff format, where +
> indicates a new insertion to the file and - represents a deletion.
>
>   * except: pass
>   * except:
>   * pass
>
> In other words, break this out on multiple lines to make it consistent
> with python best practices.

OK, I'll change it too.

> Great work.

Thanks and cheers,
-- 
¡Cuando tus fuerzas terminan, las de Dios comienzan!

francipvb committed:
just some fixes to the script.

jcsteh commented on source/appModules/eclipse.py:
```diff
@@ -17,14 +22,83 @@ def event_valueChange(self):
+class AutocompletionListView(IAccessible):
+
+	def _get_name(self):
+		return AUTOCOMPLETION_LIST_NAME
```
Does the autocomplete list in Eclipse not have a name itself? If it does, it'd be better to use that.

francipvb replied:
Removed this because as you've pointed out, it is not necessary.

jcsteh commented on source/appModules/eclipse.py:
```diff
@@ -17,14 +22,83 @@ def event_valueChange(self):
+		# Fixme: I don't know if it is the correct way to do this.
+
+		# I think that you can feel more confortable if the text editor gain the focus after the item becomes hidden.
+		if self.appModule.textEditor == None:
```
nit: Don't use "== None". Either use "is None" (faster) or just the "not" operator if there's no chance it could be False or 0. I think the latter is better in this case:

`if not self.appModule.textEditor:`

jcsteh commented on source/appModules/eclipse.py:
```diff
@@ -17,14 +22,83 @@ def event_valueChange(self):
+			self.appModule.textEditor = api.getFocusObject()
+
+		if self.appModule.selectedItem != self:
+			self.appModule.selectedItem = self
```
I assume you were seeing duplicate selection events for the same item? If not, I'd get rid of this variable.

francipvb replied:
I saw duplicate events when pressing CTRL, Shift or any other key.

jcsteh commented on source/appModules/eclipse.py:
```diff
@@ -17,14 +22,83 @@ def event_valueChange(self):
+		if self.appModule.selectedItem != self:
+			self.appModule.selectedItem = self
+			api.setFocusObject(self)
+			self.reportFocus()
```
api.setFocusObject is really for internal use (though this isn't documented properly; ug). You should use:

` eventHandler.executeEvent("gainFocus", self)`

jcsteh commented on source/appModules/eclipse.py:
```diff
@@ -17,14 +22,83 @@ def event_valueChange(self):
+			api.setFocusObject(self)
+			self.reportFocus()
+
+	def script_returnToTextEditor(self, gesture):
```
Using a script to catch dismissal of the autocomplete isn't ideal. Is there any event that gets fired when the autocomplete is dismissed? You might have to watch for "hide" events on the autocomplete list. Those are filtered out by default, so you'll need to use eventHandler.requestEvents to request them.

If a script is the only way, I'd suggest returning focus to `api.getDesktopObject().objectWithFocus()`, rather than storing the editor. This will make sure focus goes to the "real focus" even if pressing the key caused it to move somewhere other than the editor for some reason.

jcsteh commented on source/appModules/eclipse.py:
```diff
@@ -17,14 +22,83 @@ def event_valueChange(self):
+		# And clean up the selected list item because we don't need it
+		self.appModule.selectedItem = None
+
+	def script_readDocumentation(self, gesture):
```
What does this script do? What documentation? The documentation for the identifier associated with this autocomplete item? This needs a translatable docstring which will be used in input help and in the Input Gestures dialog. See the ["Scripts and Gesture Bindings" section of the Developer Gide](http://www.nvaccess.org/files/nvda/documentation/developerGuide.html#toc17) for details.

jcsteh commented on source/appModules/eclipse.py:
```diff
@@ -17,14 +22,83 @@ def event_valueChange(self):
+		self.appModule.selectedItem = None
+
+	def script_readDocumentation(self, gesture):
+		obj = api.getDesktopObject().simpleFirstChild
```
You should generally avoid using simple* properties in plugins. They can be expensive, can cause infinite recursion in some cases and might change in future NVDA versions. The Developer Guide does warn against this, though the warning could probably be stronger. Instead, figure out the correct path using the raw .firstChild, .next, etc. properties.

I'm also concerned about going from the desktop object like this. There could be cases where the first object is not even associated with your application. I assume this is some sort of floating window?

francipvb replied:
Hi,

El 17/01/2016 a las 11:49 p.m., James Teh escribió:
> In source/appModules/eclipse.py
> <https://github.com/nvaccess/nvda/pull/5680#discussion_r49958557>:
>
>> +		# We need to send the gesture to hide the autocompletion window
>> +		gesture.send()
>> +
>> +		# Retrieve the focus to the editor
>> +		if self.appModule.textEditor:
>> +			api.setFocusObject(self.appModule.textEditor)
>> +			self.appModule.textEditor.reportFocus()
>> +			
>> +			# And prevent confusing events
>> +			self.appModule.textEditor = None
>> +		
>> +		# And clean up the selected list item because we don't need it
>> +		self.appModule.selectedItem = None
>> +
>> +	def script_readDocumentation(self, gesture):
>> +		obj = api.getDesktopObject().simpleFirstChild
>
> You should generally avoid using simple* properties in plugins. They can
> be expensive, can cause infinite recursion in some cases and might
> change in future NVDA versions. The Developer Guide does warn against
> this, though the warning could probably be stronger. Instead, figure out
> the correct path using the raw .firstChild, .next, etc. properties.

OK, Thanks for the suggestion. For now, it's a basic implementation of 
the script. It works, but needs some fixes and additions.

I'll see if I can change this, because sometimes it slows down NVDA's 
performance.

> I'm also concerned about going from the desktop object like this. There
> could be cases where the first object is not even associated with your
> application. I assume this is some sort of floating window?

Yep, it's a floating window, and the Developer Guide have only basic 
documentation and I didn't read the sources. I'll try to fix it.

Thanks to all for your help.

Have a nice day.
-- 
¡Cuando tus fuerzas terminan, las de Dios comienzan!

jcsteh commented on source/appModules/eclipse.py:
```diff
@@ -17,14 +22,83 @@ def event_valueChange(self):
 
+		# I don't know why this doesn't work properly:
+		# try:
+			# if obj.role == controlTypes.ROLE_LIST and obj.simpleParent.simpleParent == api.getDesktopObject():
```
Again, avoid simple* properties. This may be why this isn't working.

jcsteh commented on source/appModules/eclipse.py:
```diff
@@ -17,14 +22,83 @@ def event_valueChange(self):
+
+		# This does, but freezes the script a lot:
+		# try:
+			# if obj.role == controlTypes.ROLE_LIST and obj.simpleParent.simpleParent == api.getDesktopObject():
```
Again, avoid simple* properties, but aside from that:

This will match any list in Eclipse that happens to be a simple grandchild of the desktop. I'm worried that this might match other lists. It'd be good to find a way to tighten this up a bit; e.g. relationship to other nearby siblings, info on the parent, etc.

jcsteh commented on source/appModules/eclipse.py:
```diff
@@ -17,14 +22,83 @@ def event_valueChange(self):
+			# pass
+
+		try:
+			if obj.role == controlTypes.ROLE_LISTITEM and obj.simpleParent.simpleParent.simpleParent == api.getDesktopObject():
```
Same as above re simple* and risk of matching other lists.

francipvb committed:
Changes:

* Improved performance, thanks to jcsteh and derekriemer.
* Changed simple* properties
* Changed `api.SetFocusObject(obj)` to `eventHandler.executeEvent(event, object)`
* Changed the readDocumentation script and it now works differently
  - Now it has a help string

Cheers,

codeofdusk commented:
Any updates on this one?

ehollig commented:
CC @feerrenrut

francipvb committed:
Merge branch 'master' into nvda-eclipse-support-improvement

feerrenrut commented on source/appModules/eclipse.py:
```diff
@@ -17,8 +25,70 @@ def event_valueChange(self):
+class AutocompletionListItem(IAccessible):
+
+	def event_selection(self):
+		# Fixme: I don't know if it is the correct way to do this.
```
This comment should be removed. If you have some concrete doubts, or cases you know are not handled add a comment to that effect.

francipvb replied:
Hello,

I don't know how to write unit tests for an app module, I didn't read about it yet.

I've downloaded eclipse to test it again.

feerrenrut reviewed (changes requested):
Could you please comment on how you have tested this? I would like to see that there has been good test coverage before we merge this in.

francipvb committed:
Merge branch 'master' into nvda-eclipse-support-improvement

francipvb committed:
Removed `event_hide` and fixed overlay class detection

francipvb committed:
Added EditableTextWithSuggestion as a base class of EclipseTextArea
Changed AutocompletionListItem to fire suggestion events
Added caret event to handle suggestion list hiding
Removed AutocompletionListView and it's related code

francipvb commented:
Hello @feerrenrut 

I've tweaked and tested this on Windows 10.

1. Open Eclipse.
2. Create a new java project.
3. Create a Main class with a standard main method.
4. Type `System.` and then press ~CTRL+Space~.
5. You will ear standard suggestion opening sound.
6. Press enter on any suggestion and check the editor.
7. Repeat step 4 and 5.
8. Press ~Right Arrow~ and you will ear the standard suggestions closing sound, like in step 6.

- [ ] This needs testing under windows 7, 8 and 8.1.

Cheers,

francipvb commented:
Another thing are the items having documentation.

1. Type `System.` and press ~CTRL+Space~.
2. Use the Arrow keys to select an item having javadoc.
3. Press ~NVDA+D~ to read the javadoc.

- [ ] Again, this needs to be tested under Windows 7, 8 and 8.1.

francipvb reviewed (commented):
Hello,

I think that most requested changes are done.

Also there are some fixes and improvements.

francipvb committed:
Change comment on `event_selection()`

feerrenrut reviewed (approved):
Thanks for the updates and the further testing. I'll incubate this shortly.

feerrenrut commented:
@francipvb Would you mind suggesting a line to add to the changes file?

francipvb commented:
Just mention something like the pr title. Note that some things needs to be tested on other supported windows versions.

Cheers,



Enviado desde mi Samsung Mobile de Claro


-------- Mensaje original --------
De: Reef Turner <notifications@github.com>
Fecha: 2/10/17 12:28 AM (GMT-03:00)
A: nvaccess/nvda <nvda@noreply.github.com>
CC: "Francisco R. Del Roio" <francipvb@hotmail.com>, Mention <mention@noreply.github.com>
Asunto: Re: [nvaccess/nvda] Improved eclipse code editing support (#5680)


@francipvb<https://github.com/francipvb> Would you mind suggesting a line to add to the changes file?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/nvaccess/nvda/pull/5680#issuecomment-333433343>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AGiczWskH8XROqwZCv5sW262UiKOV2hrks5soFhZgaJpZM4HFEz2>.

feerrenrut commented:
This seems causing automated tests to fail. In particular the test to ensure that we have translator comments. I would fix this myself, but it does not look like I have permission to push to this branch. Could you add the following before line 76 of the eclipse app module file: `# Translators: Input help mode message for the 'read documentation script' in the Eclipse app module.`

When you have done so, I'll re-incubate.

francipvb committed:
Added translators help to `readDocumentation` script help message.

dkager commented:
@feerrenrut now with NVDA master, I get this when arrowing through the auto-complete list:

```
ERROR - eventHandler.executeEvent (20:13:31.301):
error executing event: selection on <NVDAObjects.Dynamic_AutocompletionListItemListItemIAccessible object at 0x04FF1C50> with extra args of {}
Traceback (most recent call last):
  File "eventHandler.pyc", line 143, in executeEvent
  File "eventHandler.pyc", line 91, in __init__
  File "eventHandler.pyc", line 98, in next
  File "appModules\eclipse.pyc", line 91, in event_selection
AttributeError: 'Dynamic_AutocompletionListItemListItemIAccessible' object has no attribute 'event_suggestionsOpened'
```

feerrenrut commented:
Sounds like this needs more work, I'll revert the commit.

dkager commented:
Note that the list does display correctly, at least in braille, it just shows errors in the log. But then, the auto-complete list has always displayed correctly for me (after pressing Tab), so maybe that doesn't say too much.

feerrenrut commented:
I have reverted the change, considering how little time there is until the next release. I would prefer this gets worked on a little more to ensure it is not going to cause problems.

francipvb commented:
Hello,

What are you environment? Eclipse version, operating system, eclipse plugin, etc.

Normally the editor doesn’t need to lose focus to the script to work properly, but …

Cheers,

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

________________________________
De: Davy Kager <notifications@github.com>
Enviado: Wednesday, November 1, 2017 2:32:35 AM
Para: nvaccess/nvda
Cc: Francisco R. Del Roio; Mention
Asunto: Re: [nvaccess/nvda] Improved eclipse code editing support (#5680)


Note that the list does display correctly, at least in braille, it just shows errors in the log. But then, the auto-complete list has always displayed correctly for me (after pressing Tab), so maybe that doesn't say too much.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/nvaccess/nvda/pull/5680#issuecomment-340986805>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AGiczWiJaOq6UPHhbODv_vAOr5qrhI9iks5syAJzgaJpZM4HFEz2>.

dkager commented:
Windows 10 / Eclipse 4.6.3 / no special plugins or config.
